README.md
# Menu Toolkit

Menu Toolkit provides functionality to help create accessible and usable menus. Toolkit provides JavaScript functions to create the menu and it's up to you to style it and extend it.

## Menu from HTML

**Demos**

* ["Click" interaction](https://evermade.github.io/menu-toolkit/)
* ["Hover" interaction](https://evermade.github.io/menu-toolkit/hover.html)
* [Animating sub menus](https://evermade.github.io/menu-toolkit/animation.html)

Function `menuFromHTML` adds accessible enhancements and events for WordPress navigation menus (although you can use it for different menus).

Two different interaction modes are supported:

1. `click`: sub menus are opened only on click
2. `hover`: sub menus open also with hover

Things to note:

* This function only handles the `<ul>` part of your menu. You may need additional logic for example toggles or displaying mobile menu in modal
* It's recommended to have separate menu markup for mobile menu at least when hover is used (and regardless often different configuration makes sense for smaller screens)
* This is not plug-and-play library and you still need to properly style the menu

### Usage

If you’re using a bundler (such as Webpack or Rollup), you can install through npm:

```bash
npm install @evermade/menu-toolkit
```

Import the `menuFromHTML` and create a menu.

```js
import { menuFromHTML } from "@evermade/menu-toolkit";

const menuEl = document.querySelector('.js-main-menu');
if (menuEl) {
	const menu = menuFromHTML(
		menuEl,
		{
			action: 'click'
		}
	);
}
```

### Markup

In general it's recommended to wrap the menu in `<nav>` element but we'll show just the parts this function needs.

```html
<ul class="js-main-menu">
	<li><a href="/">Home</a></li>
	<li class="menu-item-has-children">
		<a href="#">Work</a>
		<ul class="sub-menu">
			<li><a href="/websites">Websites</a></li>
			<li><a href="/e-commerce">E-commerce</a></li>
			<li><a href="/apps">Apps</a></li>
	</li>
	<li class="menu-item-has-children">
		<a href="#">Services</a>
		<ul class="sub-menu">
			<li><a href="/design">Design</a></li>
			<li><a href="/development">Development</a></li>
			<li class="menu-item-has-children">
				<a href="#">Care</a>
				<ul class="sub-menu">
					<li><a href="/upkeep">Upkeep</a></li>
					<li><a href="/monitoring">Monitoring</a></li>
				</ul>
			</ul>
	</li>
	<li><a href="/blog">Blog</a></li>
</ul>
```

### Changes to markup

This functions adds toggles to open/close menus.

Before:

```html
<li class="menu-item-has-children">
	<a href="/services">Services</a>
</li>
```

After (if `shouldWrapAnchorToButton` is true):

```html
<li class="menu-item-has-children">
	<a href="/services" hidden>Services</a>
	<button aria-expanded="false" data-toggle-type="cover">
		<span>Services</span>
		<svg></svg>
	</button>
</li>
```

After (if `shouldWrapAnchorToButton` is false):

```html
<li class="menu-item-has-children">
	<a href="/services">Services</a>
	<button aria-expanded="false" data-toggle-type="icon">
		<span class="screen-reader-text">Sub menu</span>
		<svg></svg>
	</button>
</li>
```

### Functional CSS

There's much more you'll actually need but here's functional things that you need so that opening and closing sub menus work:

```CSS
[hidden] {
	display: none !important;
}

.sub-menu {
	opacity: 0;
	pointer-events: none;
	visibility: hidden;
}

.sub-menu.is-open {
	opacity: 1;
	visibility: visible;
	pointer-events: auto;
}
```

Tips and tricks:

* It's recommended not to open sub menus with `:hover` because hovering will already add `is-open` class to sub menu
* You can use selctors for toggles: `[data-toggle-type="cover"]` and `[data-toggle-type="icon"]`
* You probably want to have all the same styling for `[data-toggle-type="cover"]` than for `<a>`
* If you use animation you can do CSS animations or just transition with `.sub-menu.animate-open` and `.sub-menu.animate-close` (or whatever classes you configure)
* If you want to animate caret without delay of animation you can target it with `[data-toggle-type="cover"][aria-expanded="true"]` because aria-expanded is updated without any delay
* Remember to check that you have proper `:focus` style in place for both `<a>` and `<button>` elements

### Options

There are many settings (in object format) that can be passed as 2nd argument with following types:

```js
const menu = menuFromHTML( menuEl, {
	action: 'click',
	subMenuAnchorSelector: '.menu-item-has-children > a',
	subMenuListItemSelector: '.menu-item-has-children',
	openSubMenuClass: 'is-open',
	buttonClass: '',
	visuallyHiddenClass: 'screen-reader-text',
	expandChildMenuText: 'Sub menu',
	hoverTimeout: 750,
	buttonIcon:
		'<svg width="24" height="24" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false"><path d="M17.5 11.6L12 16l-5.5-4.4.9-1.2L12 14l4.5-3.6 1 1.2z"></path></svg>',
	shouldWrapAnchorToButton: null,
	activeListItemSelector: '.current-menu-item',
	openActiveSubMenuOnCreate: false,
	closeSubMenuOnOutsideClick: true,
	animateOpen: false,
	animateClose: false,
	animateOpenClass: 'animate-open',
	animateCloseClass: 'animate-close',
	animateOpenMaxExecutionTime: 250,
	animateCloseMaxExecutionTime: 250,
	onBeforeCreate: null,
	onAfterCreate: null,
	onBeforeOpenSubMenu: null,
	onAfterOpenSubMenu: null,
	onBeforeCloseSubMenu: null,
	onAfterCloseSubMenu: null,
});
```

#### action (string)

Either `click` or `hover`.

#### subMenuAnchorSelector (string)

Targets main `<a>` elements of menu items that have children. If using WordPress menus or WordPress-like markup, no changes needed.

Defaults to `.menu-item-has-children > a`.

#### subMenuListItemSelector (string)

Targets main `<li>`  elements of menu items that have children. If using WordPress menus or WordPress-like markup, no changes needed.

Defaults to `.menu-item-has-children`.

#### openSubMenuClass (string)

Class to be added when sub menu is open. CSS styles to show/hide sub menu should be based on this class.

Defaults to `is-open`.

#### buttonClass (string)

Class to be added to toggle buttons. Even without class you can use data-attributes as selectors.

Defaults to no added class.

#### visuallyHiddenClass (string)

Class to be added to visually hidden elements. This is used for screen readers.

Defaults to `screen-reader-text`.

#### expandChildMenuText (string)

Text to be added to toggle buttons. This is used for screen readers.

Defaults to `Sub menu`.

#### hoverTimeout (number)

Time in milliseconds to wait before closing sub menu after hover moves outside of element. This is used to prevent accidental closing of sub menus.

Defaults to `750`.

#### buttonIcon (string)

HTML string to be used as toggle button icon.

Defaults to SVG icon.

#### shouldWrapAnchorToButton (boolean | function)

Weather `<a>` elements should be wrapped inside `<button>` elements (that act as toggles). Alternatively toggle button can be separate element next to `<a>` element.

Example:

```js
shouldWrapAnchorToButton: true
```

```js
shouldWrapAnchorToButton: ( should, a, level ) => {
	return level === 1;
}
```

Defaults to wrapping when href is "#".

#### activeListItemSelector (string)

Targets active menu item. If using WordPress menus or WordPress-like markup, no changes needed.

Defaults to `.current-menu-item`.

#### openActiveSubMenuOnCreate (boolean)

If active menu item has sub menu, should it be opened on create. This should usually only be used in mobile menus or menus that are not visible on page load.

Defaults to `false`.

#### closeSubMenuOnOutsideClick (boolean)

Should sub menus be closed when clicking outside of menu. This should usually not be used on mobile menus because it may cause accidental closing of menu levels.

Defaults to `true`.

#### animateOpen (boolean)

Should sub menus be animated when opening.

Defaults to `false`.

#### animateClose (boolean)

Should sub menus be animated when closing.

Defaults to `false`.

#### animateOpenClass (string)

Class to be added when sub menu is opening. CSS styles to animate opening of sub menu should be based on this class. You can use for example animation or transitions to animate opening.

Defaults to `animate-open`.

#### animateCloseClass (string)

Class to be added when sub menu is closing. CSS styles to animate closing of sub menu should be based on this class. You can use for example animation or transitions to animate closing.

Defaults to `animate-close`.

#### animateOpenMaxExecutionTime (number)

Maximum time in milliseconds to wait for opening animation to finish. This is fallback in case animation eventhandler is not triggered.

Defaults to `250`.

#### animateCloseMaxExecutionTime (number)

Maximum time in milliseconds to wait for closing animation to finish. This is fallback in case animation eventhandler is not triggered.

Defaults to `250`.

#### onBeforeCreate (function)

Callback function to be called before menu is created. This is called only once. If you need to modify markup before menu is created, this is the place to do it.

Example:

```js
onBeforeCreate: ( menu ) => {
	// wrap all anchors to span
	const anchors = menu.querySelectorAll('a');
	anchors.forEach((anchor) => {
		const span = document.createElement('span');
		anchor.parentNode.insertBefore(span, anchor);
		span.appendChild(anchor);
	});
}
```

Defaults to `null`.

#### onAfterCreate (function)

Callback function to be called after menu is created. This is called only once. If you need to modify markup after menu is created, this is the place to do it.

Example:

```js
onAfterCreate: ( menu ) => {}
```

Defaults to `null`.

#### onBeforeOpenSubMenu (function)

Callback function to be called before sub menu is opened. This is called every time sub menu is opened. If you need to modify markup or styling in a way that could cause flickering after sub menu is opened, this is the place to do it.

Example:

```js
onBeforeOpenSubMenu: (ul) => {
	// check that sub-menu fits on right side of parent
	if (ul.getBoundingClientRect().right > (window.innerWidth || document.documentElement.clientWidth)) {
		ul.classList.add('is-out-of-bounds');
	} else {
		ul.classList.remove('is-out-of-bounds');
	}
}
```

Defaults to `null`.

#### onAfterOpenSubMenu (function)

Callback function to be called after sub menu is opened. This is called every time sub menu is opened.

Example:

```js
onAfterOpenSubMenu: (ul) => {}
```

Defaults to `null`.

#### onBeforeCloseSubMenu (function)

Callback function to be called before sub menu is closed. This is called every time sub menu is closed.

Example:

```js
onBeforeCloseSubMenu: (ul) => {}
```

Defaults to `null`.

#### onAfterCloseSubMenu (function)

Callback function to be called after sub menu is closed. This is called every time sub menu is closed. If you need to modify markup or styling in a way that could cause flickering before sub menu is closed, this is the place to do it.

Example:

```js
onAfterCloseSubMenu: (ul) => {
	// remove out of bounds class
	ul.classList.remove('is-out-of-bounds');
}
```

Defaults to `null`.

## Acknowledgements

This library has taken a lot from [MEOM/navigation](https://github.com/MEOM/navigation). Although we have a different approach to many things, it has been a great source of inspiration and ideas.

## Development

Install tools `npm install` and build `npm run build` or develop with `npm run watch`.

Releasing new version:

* Update version in `package.json`
* Commit to master
* Set tag with version number to git
* Create new release in GitHub
* NPM package is automatically published from GitHub


changelog.md
# Changelog

## Version 1.0.0 - 2023-08-24

- Initial release


rollup.config.js
import { terser } from 'rollup-plugin-terser';
import pkg from './package.json';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import typescript from '@rollup/plugin-typescript';

const umd = { format: 'umd', name: 'Menu', exports: 'named' };
const es = { format: 'es' };
const minify = {
	plugins: [terser()],
	banner: () => `/*!  ${pkg.name} ${pkg.version} */`,
};

export default {
	input: 'src/index.ts',
	output: [
		// Main files
		{ file: 'dist/index.js', ...umd },
		{ file: 'dist/index.esm.js', ...es },
		{ file: 'docs/dist/menu-toolkit.esm.js', ...es },
		// Minified versions
		{ file: 'dist/index.min.js', ...umd, ...minify },
		{ file: 'dist/index.esm.min.js', ...es, ...minify },
	],
	plugins: [
		typescript(),
		nodeResolve(),
		commonjs({ include: 'node_modules/**' }),
	],
};


demo/assets/demo.js
import { menuFromHTML } from '../dist/menu-toolkit.esm.js';

(function () {
	const init = () => {
		const demoId = document.body.getAttribute('data-demo');

		switch (demoId) {
			case 'click':
				demoClick();
				break;
			case 'hover':
				demoHover();
				break;
			case 'animation':
				demoAnimation();
				break;
			default:
				throw new Error(`No matching demo found. Demo ID: ${demoId}`);
		}
	};

	const demoClick = () => {
		menuFromHTML(document.querySelector('.js-menu-toolkit'), {
			action: 'click',
		});
	};

	const demoHover = () => {
		menuFromHTML(document.querySelector('.js-menu-toolkit'), {
			action: 'hover',
		});
	};

	const demoAnimation = () => {
		menuFromHTML(document.querySelector('.js-menu-toolkit'), {
			action: 'hover',
			animateOpen: true, // class 'animate-open'
			animateClose: true, // class 'animate-close'
		});
	};

	init();
})();


src/utils/animate.ts
export type AnimateArgs = {
	element: HTMLElement;
	animationClass: string;
	onReady?: (element: HTMLElement, animationClass: string) => void;
	maxExecutionTime?: number;
};

export default function animate(options: AnimateArgs) {
	const { element, animationClass, onReady, maxExecutionTime } = options;

	// Apply the animation.
	element.classList.add(animationClass);

	let isReadyCallbackRun = false;

	function endAnimation() {
		element.removeEventListener('animationend', endAnimation, false);
		if (typeof onReady === 'function') {
			isReadyCallbackRun = true;
			onReady(element, animationClass);
		}
	}

	// Detect when the animation ends.
	element.addEventListener('animationend', endAnimation, false);

	// Run setTimeout to ensure the animation ends.
	if (typeof onReady === 'function' && maxExecutionTime) {
		setTimeout(() => {
			if (!isReadyCallbackRun) {
				isReadyCallbackRun = true;
				onReady(element, animationClass);
				element.removeEventListener(
					'animationend',
					endAnimation,
					false
				);
			}
		}, maxExecutionTime);
	}
}


src/menuFromHTML.ts
import animate from './utils/animate';

type El = HTMLElement;

export type MenuActionType = 'click' | 'hover';

export type MenuSettings = {
	action: MenuActionType;
	subMenuAnchorSelector: string;
	subMenuListItemSelector: string;
	openSubMenuClass: string;
	buttonClass: ((el: El, subMenu: El, level: number) => string) | string;
	visuallyHiddenClass: string;
	expandChildMenuText: string;
	hoverTimeout: number;
	buttonIcon: ((el: El, subMenu: El, level: number) => string) | string;
	shouldWrapAnchorToButton:
		| ((should: boolean, anchor: El, level: number) => boolean)
		| boolean
		| null;
	activeListItemSelector: string;
	openActiveSubMenuOnCreate: boolean;
	closeSubMenuOnOutsideClick: boolean;
	animateOpen: boolean;
	animateClose: boolean;
	animateOpenClass: string;
	animateCloseClass: string;
	animateOpenMaxExecutionTime: number;
	animateCloseMaxExecutionTime: number;
	onBeforeCreate: (() => void) | null;
	onAfterCreate: (() => void) | null;
	onBeforeOpenSubMenu: ((ul: El, event?: Event) => void) | null;
	onAfterOpenSubMenu: ((ul: El, event?: Event) => void) | null;
	onBeforeCloseSubMenu: ((ul: El, event?: Event) => void) | null;
	onAfterCloseSubMenu: ((ul: El, event?: Event) => void) | null;
};

export type MenuSettingsArgs = Partial<MenuSettings>;

const TAB_KEY = 'Tab';
const ESC_KEY = 'Escape';

const FOCUSABLE_ELEMENTS = `
a[href]:not([hidden]),
area[href],
input:not([disabled]),
select:not([disabled]),
textarea:not([disabled]),
button:not([disabled])
`;

export default function menuFromHTML(
	element: El,
	options?: MenuSettingsArgs | null
) {
	// Elements.
	const $ulRoot: El = element;

	let hoverTimeoutIndex: { id: ReturnType<typeof setTimeout>; ul: El }[] = [];

	// Validate element.
	if (!(element instanceof HTMLElement)) {
		throw new Error('Menu element must be an valid HTMLElement');
	}

	const defaults: MenuSettings = {
		action: 'click',
		subMenuAnchorSelector: '.menu-item-has-children > a',
		subMenuListItemSelector: '.menu-item-has-children',
		openSubMenuClass: 'is-open',
		buttonClass: '',
		visuallyHiddenClass: 'screen-reader-text',
		expandChildMenuText: 'Sub menu',
		hoverTimeout: 750,
		buttonIcon:
			'<svg width="24" height="24" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false"><path d="M17.5 11.6L12 16l-5.5-4.4.9-1.2L12 14l4.5-3.6 1 1.2z"></path></svg>',
		shouldWrapAnchorToButton: null,
		activeListItemSelector: '.current-menu-item',
		openActiveSubMenuOnCreate: false, // should be used only for mobile menus
		closeSubMenuOnOutsideClick: true,
		animateOpen: false,
		animateClose: false,
		animateOpenClass: 'animate-open',
		animateCloseClass: 'animate-close',
		animateOpenMaxExecutionTime: 250,
		animateCloseMaxExecutionTime: 250,
		onBeforeCreate: null,
		onAfterCreate: null,
		onBeforeOpenSubMenu: null,
		onAfterOpenSubMenu: null,
		onBeforeCloseSubMenu: null,
		onAfterCloseSubMenu: null,
	};

	// Merge options to defaults.
	const settings: MenuSettings = { ...defaults, ...options };

	const $subMenuAnchors: El[] = Array.from(
		$ulRoot.querySelectorAll(settings.subMenuAnchorSelector)
	) as El[];
	const $subMenuListItems: El[] = Array.from(
		$ulRoot.querySelectorAll(settings.subMenuListItemSelector)
	) as El[];

	/**
	 * Create the menu.
	 */
	const create = () => {
		/**
		 * Called before the component is initialized.
		 */
		if (typeof settings.onBeforeCreate === 'function') {
			settings.onBeforeCreate();
		}

		// Set data value for nav element. This is for targeting without a class name.
		$ulRoot.setAttribute('data-menu', 'root');

		// Setup sub menu toggle buttons.
		let uniqueId = 1;
		$subMenuAnchors.forEach((a, index) => {
			const button = document.createElement('button');
			button.setAttribute('data-menu', 'sub-toggle');
			button.setAttribute('aria-expanded', 'false');

			// Add level as data attribute by counting the number of parents.
			let level = 1;
			let parent = a.parentElement;
			while (parent && parent !== $ulRoot) {
				if (parent.tagName === 'UL') {
					level++;
				}
				parent = parent.parentElement;
			}
			button.setAttribute('data-menu-level', `${level}`);

			while (document.getElementById(`sub-menu-${uniqueId}-${index}`)) {
				uniqueId++;
			}
			button.setAttribute(
				'aria-controls',
				`sub-menu-${uniqueId}-${index}`
			);

			// Add matching id for next sub-sub-menu.
			const ul = a.closest('li')?.querySelector('ul') as El;
			if (ul) {
				ul.id = `sub-menu-${uniqueId}-${index}`;
				ul.setAttribute('data-menu', 'sub-menu');
			}

			button.className = getButtonClass(ul, level);
			button.type = 'button';

			const buttonIcon = getButtonIcon(ul, level);

			if (shouldWrapAnchorToButton(a, level)) {
				button.innerHTML = `<span>${a.textContent}</span>${buttonIcon}`;
				button.setAttribute('data-toggle-type', 'cover');
				a.after(button);
				// Hide link in JS to avoid cumulative layout shift (CLS).
				a.setAttribute('hidden', '');
			} else {
				button.innerHTML = `<span class="${settings.visuallyHiddenClass}">${settings.expandChildMenuText}</span>${buttonIcon}`;
				button.setAttribute('data-toggle-type', 'icon');
				a.after(button);
			}
		});

		// Open active sub menu on create.
		if (settings.openActiveSubMenuOnCreate) {
			openActiveSubMenu();
		}

		// Set event listeners.
		$ulRoot.addEventListener('click', handleSubMenu, false);
		$ulRoot.addEventListener('keydown', handleCloseSubMenu, false);

		if (settings.closeSubMenuOnOutsideClick) {
			document.addEventListener('click', handleOutsideClick, false);
		}

		if (settings.action === 'hover') {
			$subMenuListItems.forEach((li) => {
				li.addEventListener('mouseenter', handleMouseEnter, false);
				li.addEventListener('mouseleave', handleMouseLeave, false);
			});
		}

		/**
		 * Called after the component is initialized.
		 */
		if (typeof settings.onAfterCreate === 'function') {
			settings.onAfterCreate();
		}

		$ulRoot.classList.add('is-ready');
	};

	/**
	 * Handle sub menu toggle button click.
	 *
	 * @param event
	 */
	const handleSubMenu = (event: Event) => {
		const target = event.target as El;

		const button = target.closest('[data-menu="sub-toggle"]');
		if (!button) {
			return;
		}

		const ul = $ulRoot.querySelector(
			`#${button.getAttribute('aria-controls')}`
		) as El;
		if (!ul) {
			return;
		}

		if (isOpenSubMenu(ul)) {
			closeSubMenu(ul, event);
		} else {
			// Get all sub menus that are not the closest sub menu or its parents.
			const subMenusToClose = getOpenSubMenus().filter((subMenu) => {
				return subMenu !== ul && !subMenu.contains(ul);
			});
			if (subMenusToClose.length) {
				subMenusToClose.forEach((subMenu: El) => {
					closeSubMenu(subMenu);
				});
			}
			openSubMenu(ul, event);
		}
	};

	/**
	 * Handle closing sub-navs when clicking outside of nav.
	 *
	 * @param event
	 */
	const handleOutsideClick = (event: MouseEvent) => {
		const target = event.target as El;
		// Bail if clicking inside the nav.
		if (target.closest('[data-menu="root"]')) {
			return;
		}

		closeAllSubMenus();
	};

	/**
	 * Handle close sub menu with Tab or Esc key.
	 *
	 * @param event
	 */
	const handleCloseSubMenu = (event: KeyboardEvent) => {
		const target = event.target as El;

		const closestSubButtonToChild = target
			.closest('[data-menu="sub-menu"]')
			?.parentElement?.querySelector('[data-menu="sub-toggle"]');
		const closestSubButtonToParent = target.closest(
			'[data-menu="sub-toggle"]'
		);
		let closestSubButton;
		if (target.getAttribute('data-menu') === 'sub-toggle') {
			closestSubButton = target;
		} else if (closestSubButtonToChild) {
			closestSubButton = closestSubButtonToChild;
		} else {
			closestSubButton = closestSubButtonToParent;
		}

		const closestListItem = closestSubButton
			? closestSubButton.closest('li')
			: target.closest('li');
		if (!closestListItem) {
			return;
		}

		// Focusable elements.
		const relevantFocusableElements = getFocusableElements(closestListItem);
		const lastFocusableElement =
			relevantFocusableElements[relevantFocusableElements.length - 1];

		// Last Tab closes sub-menu.
		if (
			TAB_KEY === event.key &&
			!event.shiftKey &&
			event.target === lastFocusableElement
		) {
			const closableSubMenus = getOpenSubMenus().filter((subMenu: El) => {
				const subMenuFocusableElements = getFocusableElements(subMenu);
				const isLastFocusableElement =
					subMenuFocusableElements[
						subMenuFocusableElements.length - 1
					] === event.target;
				if (isLastFocusableElement) {
					return true;
				}
				if (!subMenu.contains(closestListItem)) {
					return true;
				}
				return false;
			});
			if (closableSubMenus.length) {
				closableSubMenus.forEach((subMenu) => {
					closeSubMenu(subMenu);
				});
			}
		}

		// First Shift + Tab closes sub-menu.
		if (
			closestSubButton &&
			TAB_KEY === event.key &&
			event.shiftKey &&
			event.target === closestSubButton
		) {
			const closableSubMenus = getOpenSubMenus().filter((subMenu) => {
				return !subMenu.contains(closestListItem);
			});
			if (closableSubMenus.length) {
				closableSubMenus.forEach((subMenu) => {
					closeSubMenu(subMenu);
				});
			}
		}

		// Esc closes sub-menu.
		if (ESC_KEY === event.key) {
			const closestParentListItem = target
				.closest('[data-menu="sub-menu"]')
				?.closest('li');
			if (closestParentListItem) {
				const focusableElements = getFocusableElements(
					closestParentListItem
				);
				if (focusableElements.length) {
					focusableElements[0].focus();
				}
				const closestParentSubMenu =
					closestParentListItem.querySelector(
						'[data-menu="sub-menu"]'
					) as El;
				if (closestParentSubMenu) {
					// if is open => close
					if (isOpenSubMenu(closestParentSubMenu)) {
						closeSubMenu(closestParentSubMenu);
					}
				}
			}
		}
	};

	/**
	 * Open sub menu.
	 *
	 * @param ul
	 * @param event
	 */
	const openSubMenu = (ul: El, event?: Event) => {
		// Bail if already open.
		if (isOpenSubMenu(ul)) {
			return;
		}

		/**
		 * Called before the sub menu is opened.
		 */
		if (typeof settings.onBeforeOpenSubMenu === 'function') {
			settings.onBeforeOpenSubMenu(ul, event);
		}

		const button = $ulRoot.querySelector(`[aria-controls="${ul.id}"]`);

		if (button) {
			button.setAttribute('aria-expanded', 'true');
		}

		// Remove animation class if exists.
		if (settings.animateCloseClass) {
			ul.classList.remove(settings.animateCloseClass);
		}

		ul.classList.add(settings.openSubMenuClass);

		const openReady = () => {
			// Remove animation class if exists.
			if (settings.animateOpenClass) {
				ul.classList.remove(settings.animateOpenClass);
			}
			/**
			 * Called after the sub menu is opened.
			 */
			if (typeof settings.onAfterOpenSubMenu === 'function') {
				settings.onAfterOpenSubMenu(ul, event);
			}
		};

		if (settings.animateOpen) {
			animate({
				element: ul,
				animationClass: settings.animateOpenClass,
				onReady: () => {
					openReady();
				},
				maxExecutionTime: settings.animateOpenMaxExecutionTime,
			});
		} else {
			openReady();
		}
	};

	/**
	 * Close sub menu.
	 *
	 * @param ul
	 * @param event
	 */
	const closeSubMenu = (ul: El, event?: Event) => {
		// Bail if already closed.
		if (!isOpenSubMenu(ul)) {
			return;
		}

		const button = $ulRoot.querySelector(`[aria-controls="${ul.id}"]`);

		if (button) {
			button.setAttribute('aria-expanded', 'false');
		}

		const closeReady = () => {
			ul.classList.remove(settings.openSubMenuClass);

			// Remove animation class if exists.
			if (settings.animateCloseClass) {
				ul.classList.remove(settings.animateCloseClass);
			}

			/**
			 * Called after the sub menu is closed.
			 */
			if (typeof settings.onAfterCloseSubMenu === 'function') {
				settings.onAfterCloseSubMenu(ul, event);
			}
		};

		if (settings.animateClose) {
			animate({
				element: ul,
				animationClass: settings.animateCloseClass,
				onReady: () => {
					closeReady();
				},
				maxExecutionTime: settings.animateCloseMaxExecutionTime,
			});
		} else {
			closeReady();
		}
	};

	/**
	 * Handle mouse enter list item.
	 *
	 * @param event
	 */
	const handleMouseEnter = (event: MouseEvent) => {
		const currentTarget = event.currentTarget as El;

		const ul = currentTarget.querySelector('[data-menu="sub-menu"]') as El;

		// Clear any previously set closing timeout for this element or its parents.
		hoverTimeoutIndex = hoverTimeoutIndex.filter((timeout) => {
			if (timeout.ul === ul || timeout.ul.contains(ul)) {
				clearTimeout(timeout.id);
				return false;
			}
			// Remove this sub menu from the array.
			return true;
		});

		// close any open sub menus that are not parents or children of the current target
		const subMenusToClose = getOpenSubMenus().filter((subMenu) => {
			return (
				subMenu !== ul && !subMenu.contains(ul) && !ul.contains(subMenu)
			);
		});
		subMenusToClose.forEach((subMenu) => {
			closeSubMenu(subMenu, event);
		});

		if (ul) {
			openSubMenu(ul, event);
		}
	};

	/**
	 * Handle mouse leave list item.
	 *
	 * @param event
	 */
	const handleMouseLeave = (event: MouseEvent) => {
		const currentTarget = event.currentTarget as El;
		const ul = currentTarget.querySelector('[data-menu="sub-menu"]') as El;
		if (ul) {
			hoverTimeoutIndex = hoverTimeoutIndex.filter((timeout) => {
				// Clear any previously set closing timeout for this sub menu.
				if (timeout.ul === ul) {
					clearTimeout(timeout.id);
					return false;
				}
				return true;
			});
			// Set a new closing timeout for this sub menu.
			hoverTimeoutIndex.push({
				id: setTimeout(
					(refUl) => {
						closeSubMenu(refUl);
					},
					settings.hoverTimeout,
					ul
				),
				ul,
			});
		}
	};

	/*
	 * Is open sub menu.
	 */
	const isOpenSubMenu = (ul: El): boolean => {
		return ul.classList.contains(settings.openSubMenuClass);
	};

	/**
	 * Get open sub menus.
	 */
	const getOpenSubMenus = (): Array<El> => {
		return Array.from(
			$ulRoot.querySelectorAll(
				`.${settings.openSubMenuClass}[data-menu="sub-menu"]`
			)
		) as El[];
	};

	/**
	 * Get focusable elements-
	 *
	 * @param targetEl
	 */
	const getFocusableElements = (targetEl: El): Array<El> => {
		const focusableElements = Array.from(
			targetEl.querySelectorAll(FOCUSABLE_ELEMENTS)
		) as El[];
		// filter out all that have parent sub-menu that is not open
		const focusableElementsFiltered = focusableElements.filter(
			(focusableElement) => {
				const closestParentSubMenu = focusableElement.closest(
					'[data-menu="sub-menu"]'
				) as El;
				if (!closestParentSubMenu) {
					return true;
				}
				return isOpenSubMenu(closestParentSubMenu);
			}
		);
		return focusableElementsFiltered;
	};

	/**
	 * Close all sub menus.
	 */
	const closeAllSubMenus = () => {
		const openSubMenus = getOpenSubMenus();
		if (openSubMenus.length) {
			openSubMenus.forEach((subMenu) => {
				closeSubMenu(subMenu);
			});
		}
	};

	/**
	 * Open active sub menu.
	 */
	const openActiveSubMenu = () => {
		const liActive = $ulRoot.querySelector(
			settings.activeListItemSelector
		) as El;
		if (liActive) {
			let ulActive = liActive.closest('[data-menu="sub-menu"]') as El;
			if (!ulActive) {
				ulActive = liActive.querySelector(
					'[data-menu="sub-menu"]'
				) as El;
			}
			if (ulActive) {
				const ulActiveParents = [];
				let parent = ulActive.parentElement;
				while (parent && parent !== $ulRoot) {
					if (parent.tagName === 'UL') {
						ulActiveParents.push(parent);
					}
					parent = parent.parentElement;
				}
				const ulActiveEls = [...ulActiveParents, ulActive];
				ulActiveEls.forEach((ul) => {
					openSubMenu(ul);
				});
			}
		}
	};

	/**
	 * Should wrap anchor in button.
	 *
	 * @param a
	 * @param level
	 */
	const shouldWrapAnchorToButton = (a: El, level: number): boolean => {
		const href = a.getAttribute('href');
		let should = true;
		if (href && href !== '#') {
			should = false;
		}
		if (typeof settings.shouldWrapAnchorToButton === 'boolean') {
			should = settings.shouldWrapAnchorToButton;
		} else if (typeof settings.shouldWrapAnchorToButton === 'function') {
			should = settings.shouldWrapAnchorToButton(should, a, level);
		}
		return should;
	};

	/**
	 * Get button class.
	 *
	 * @param ul
	 * @param level
	 */
	const getButtonClass = (ul: El, level: number): string => {
		return typeof settings.buttonClass === 'function'
			? settings.buttonClass($ulRoot, ul, level)
			: settings.buttonClass;
	};

	/**
	 * Get button icon.
	 *
	 * @param ul
	 * @param level
	 */
	const getButtonIcon = (ul: El, level: number): string => {
		return typeof settings.buttonIcon === 'function'
			? settings.buttonIcon($ulRoot, ul, level)
			: settings.buttonIcon;
	};

	create();

	return {
		openSubMenu,
		closeSubMenu,
		closeAllSubMenus,
		openActiveSubMenu,
		getOpenSubMenus,
		getButtonClass,
		getButtonIcon,
	};
}
